File: java/PhaseGateOne/PhonebookApp/Java/PhoneBookApp.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class PhoneBookApp {
    private static Map<String, String> contacts = new HashMap<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        boolean userIsOnApp = true;
        while (userIsOnApp) {
            System.out.println("Phone Book App");
            System.out.println("1. Add contact");
            System.out.println("2. Remove contact");
            System.out.println("3. Find contact by phone number");
            System.out.println("4. Find contact by first name");
            System.out.println("5. Find contact by last name");
            System.out.println("6. Edit contact");
            System.out.println("7. Display contacts");
            System.out.println("8. Quit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            scanner.nextLine(); // Consume newline left-over

            switch (choice) {
                case 1:
                    addContact();
                    break;
                case 2:
                    removeContact();
                    break;
                case 3:
                    findContactByPhoneNumber();
                    break;
                case 4:
                    findContactByFirstName();
                    break;
                case 5:
                    findContactByLastName();
                    break;
                case 6:
                    editContact();
                    break;
                case 7:
                    displayContacts();
                    break;
                case 8:
                    System.out.println("Goodbye");
                    userIsOnApp = false;
                    break;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }

    private static void addContact() {
        System.out.print("Enter contact name: ");
        String name = scanner.nextLine();
        System.out.print("Enter phone number: ");
        String phoneNumber = scanner.nextLine();
        contacts.put(name, phoneNumber);
    }

    private static void removeContact() {
        System.out.print("Enter contact name: ");
        String name = scanner.nextLine();
        if (contacts.containsKey(name)) {
            contacts.remove(name);
        } else {
            System.out.println("Contact not found");
        }
    }

    private static void findContactByPhoneNumber() {
        System.out.print("Enter phone number: ");
        String phoneNumber = scanner.nextLine();
        for (Map.Entry<String, String> entry : contacts.entrySet()) {
            if (entry.getValue().equals(phoneNumber)) {
                System.out.println("Contact name: " + entry.getKey());
                return;
            }
        }
        System.out.println("Contact not found");
    }

    private static void findContactByFirstName() {
        System.out.print("Enter first name: ");
        String firstName = scanner.nextLine().toLowerCase();
        for (Map.Entry<String, String> entry : contacts.entrySet()) {
            String[] names = entry.getKey().split(" ");
            if (names[0].toLowerCase().equals(firstName)) {
                System.out.println("Contact name: " + entry.getKey() + ", Phone number: " + entry.getValue());
                return;
            }
        }
        System.out.println("Contact not found");
    }

    private static void findContactByLastName() {
        System.out.print("Enter last name: ");
        String lastName = scanner.nextLine().toLowerCase();
        for (Map.Entry<String, String> entry : contacts.entrySet()) {
            String[] names = entry.getKey().split(" ");
            if (names[names.length - 1].toLowerCase().equals(lastName)) {
                System.out.println("Contact name: " + entry.getKey() + ", Phone number: " + entry.getValue());
                return;
            }
        }
        System.out.println("Contact not found");
    }

    private static void editContact() {
        System.out.print("Enter contact name: ");
        String name = scanner.nextLine();
        if (contacts.containsKey(name)) {
            System.out.print("Enter new phone number: ");
            String newPhoneNumber = scanner.nextLine();
            contacts.put(name, newPhoneNumber);
        } else {
            System.out.println("Contact not found");
        }
    }

    private static void displayContacts() {
        System.out.println("Contacts:");
        for (Map.Entry<String, String> entry : contacts.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }
    }
}

File: java/PhaseGateOne/PhonebookApp/Python/phonebook_app.py
contacts = {}

def add_contact():
    name = input("Enter contact name: ")
    phone_number = input("Enter phone number: ")
    contacts[name] = phone_number

def remove_contact():
    name = input("Enter contact name: ")
    if name in contacts:
        del contacts[name]
    else:
        print("Contact not found")

def find_contact_by_phone_number():
    phone_number = input("Enter phone number: ")
    for name, number in contacts.items():
        if number == phone_number:
            print(f"Contact name: {name}")
            return
    print("Contact not found")

def find_contact_by_first_name():
    first_name = input("Enter first name: ")
    for name, phone_number in contacts.items():
        if name.split()[0].lower() == first_name.lower():
            print(f"Contact name: {name}, Phone number: {phone_number}")
            return
    print("Contact not found")

def find_contact_by_last_name():
    last_name = input("Enter last name: ")
    for name, phone_number in contacts.items():
        if name.split()[-1].lower() == last_name.lower():
            print(f"Contact name: {name}, Phone number: {phone_number}")
            return
    print("Contact not found")

def edit_contact():
    name = input("Enter contact name: ")
    if name in contacts:
        new_phone_number = input("Enter new phone number: ")
        contacts[name] = new_phone_number
    else:
        print("Contact not found")

def display_contacts():
    print("Contacts:")
    for name, phone_number in contacts.items():
        print(f"{name}: {phone_number}")

def main():
    user_is_on_app = True
    while user_is_on_app:
        print("Phone Book App")
        print("1. Add contact")
        print("2. Remove contact")
        print("3. Find contact by phone number")
        print("4. Find contact by first name")
        print("5. Find contact by last name")
        print("6. Edit contact")
        print("7. Display contacts")
        print("8. Quit")
        choice = input("Enter your choice: ")
        match choice:
            case "1":
                add_contact()
            case "2":
                remove_contact()
            case "3":
                find_contact_by_phone_number()
            case "4":
                find_contact_by_first_name()
            case "5":
                find_contact_by_last_name()
            case "6":
                edit_contact()
            case "7":
                display_contacts()
            case "8":
                print("Goodbye")
                user_is_on_app = False
            case _:
                print("Invalid choice. Please try again.")


main()
